[ENH] For classification: Adding InceptionTime, CNN and ResNet along with other classifiers and removing LSTM-FCN

hadifawaz1999 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/sktime/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs 

Added InceptionTime for classification as mentioned in #106 with the option to use [custom filters](https://maxime-devanne.com/pages/HCCF-4-tsc/).

Fixes #106 
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.

 I made the models CNN, ResNet and InceptionTime parametrized so the users can change in the architecture how they like, default parameters are taken from [here](https://github.com/hfawaz/dl-4-tsc). Some comments about some code lines in CNN are added, noticed some questions about them.

I started by adding missing classifiers from the original [review](https://github.com/hfawaz/dl-4-tsc) such as Encoder. More classifiers will be added on the same branch.

I removed LSTM-FCN classifier, given the issues about monitoring on the test set. The model as well has a feature of applying LSTM on 1 time stamp.
For more information, the following issues should be addressed:

1. [monitoring the val loss](https://github.com/titu1994/LSTM-FCN/issues/7#issuecomment-423917477)
2. [lstm on 1 time stamp](https://github.com/titu1994/LSTM-FCN/issues/4#issue-323439220)
<!--
A clear and concise description of what you have implemented.
-->

#### Does your contribution introduce a new dependency? If yes, which one?

yes [tensorflow_addons](https://github.com/tensorflow/addons) for the Encoder classifier.

<!--
If your contribution does add a new hard dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum.
-->

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Did you add any tests for the change?

<!-- This section is useful if you have added a test in addition to the existing ones. This will ensure that further changes to these files won't introduce the same kind of bug. It is considered good practice to add tests with newly added code to enforce the fact that the code actually works. This will reduce the chance of introducing logical bugs.
-->

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.


<!--
Thanks for contributing!
-->

TonyBagnall reviewed (changes requested):
this is great, but needs a bit of work to conform. Firstly, InceptionTime should inherit from BaseClassifier and implement    _fit/_predict not fit/predict. Secondly, for scikit learn compliance, you should not change the value stored in variables passes as arguments to the constructor, you should store in a new private variable, so rather than this in the constructor

     if isinstance(kernel_sizes, list):
            self.kernel_sizes = kernel_sizes
        else:
            self.kernel_sizes = [kernel_sizes] * n_layers

you would do this in the constructor
self.kernel_sizes = kernel_sizes

then in fit you do this
     if isinstance(self.kernel_sizes, list):
            self._kernel_sizes = self.kernel_sizes
        else:
            self._kernel_sizes = [self.kernel_sizes] * n_layers

its for things like CV and definitely a scikit thing not ours :) At least as I understand it, someone else will tell me if thats incorrect.

Only fields in the child class need to be listed in the doc string/

hadifawaz1999 commented on sktime/classification/deep_learning/cnn.py:
```diff
@@ -20,34 +20,48 @@ class CNNClassifier(BaseDeepClassifier):
 
     Parameters
     ----------
-    should inherited fields be listed here?
```
Thank you for you comments, I did the necessary changes:

1. InceptionTime now inherits from BaseClassifier with _fit _predict and _predict_proba as functions instead of fit predict and predict_proba
2. Took care of the change in argument variables changing and set them as private variables before assignment

TonyBagnall reviewed (changes requested):
great, a few more points, some are improvements to existing code, but since we are here :)
1. a couple of formatting issues to sort out, failing code quality.  
2. the next ask is correctness testing. This is put in place to make sure the algorithm doesn't get changed without people noticing. Look in classification/tests/ for the model for it, ask on slack if its not clear
3. The access format in the existing deep learning is different to the others, so may as well change here. If you look in, for example, interval based, the classifiers are private and accessed through the init method. Good to have the same pattern in all
4. why are there separate files lstmfcn and lstfcm_layers in networks? coould the layers go in the same file?

hadifawaz1999 commented:
Okay I will go through all that. For now I took care of some code quality errors, not sure why locally they did not appear before.
For lstm fcn, I think it does not matter, because I actually deleted it in my branch if am not mistaken and added in the comment of the PR why I removed it. Am not sure why it was implemented as lstmfcn and lstmfcn_layer independently though.

MatthewMiddlehurst reviewed (changes requested):
I think the InceptionTime network is missing some dependency tags i.e. `_tags = {"python_dependencies": "tensorflow"}` from CNN.

![image](https://user-images.githubusercontent.com/25731235/223192473-173695bd-0816-4beb-97ab-6679bd049433.png)

MatthewMiddlehurst commented on sktime/classification/deep_learning/encoder.py:
```diff
@@ -0,0 +1,272 @@
+        """
+        from sktime.utils.validation._dependencies import _check_soft_dependencies
+
+        param1 = {
+            "n_epochs": 10,
+            "batch_size": 4,
+            "avg_pool_size": 4,
+        }
```
"avg_pool_size" is not a parameter which is causing one of the test failures.

MatthewMiddlehurst commented:
Thanks for contributing! There may be a couple more comments/requests, but best to get the current CI failures out of the way.

Feel free to ask if there are any questions.

hadifawaz1999 commented on sktime/classification/deep_learning/cnn.py:
```diff
@@ -20,34 +20,51 @@ class CNNClassifier(BaseDeepClassifier):
+        if not a list, the same condition is used for all layers
+    random_state    : int, default = 0
+        seed to any needed random actions
+    nb_epochs       : int, default = 2000
         the number of epochs to train the model
```
Actually for this parameter, its easier if my code adapts the notebooks, so which one would you want me to use? I will use it through all my branch.

TonyBagnall replied:
I would prefer n_epochs I think, since its that already

hadifawaz1999 replied:
okay i will make the change, thanks

TonyBagnall commented on sktime/classification/deep_learning/inception_time.py:
```diff
@@ -0,0 +1,673 @@
+class InceptionTimeClassifier(BaseClassifier):
+    """InceptionTime ensemble classifier.
+
+    Ensemble of IndividualInceptionTimeClassifiers, as desribed in [1].
```
described

TonyBagnall commented on sktime/classification/deep_learning/inception_time.py:
```diff
@@ -0,0 +1,673 @@
+        use_custom_filters  : bool, default = True,
+            condition on wether or not to use custom
+            filters in the first inception module
+
```
no blank line

TonyBagnall commented on sktime/classification/deep_learning/inception_time.py:
```diff
@@ -0,0 +1,673 @@
+    https://github.com/hfawaz/InceptionTime/blob/master/classifiers/inception.py
+
+    and Ismail-Fawaz et al.
+    https://github.com/MSD-IRIMAS/CF-4-TSC
```
an Example would be good

hadifawaz1999 replied:
an Example was added with fit and predict functions used

TonyBagnall commented on sktime/classification/deep_learning/inception_time.py:
```diff
@@ -0,0 +1,673 @@
+
+        self.classifers_ = []
+
+        # super(InceptionTimeClassifier, self).__init__()
```
how come this is commented out?

TonyBagnall commented on sktime/classification/deep_learning/inception_time.py:
```diff
@@ -0,0 +1,673 @@
+            cls.fit(X, y)
+            self.classifers_.append(cls)
+
+            self.classes_ = cls.classes_
```
you dont need to set this inside the loop, in fact its set in base class fit()

TonyBagnall commented on sktime/classification/deep_learning/inception_time.py:
```diff
@@ -0,0 +1,673 @@
+        Arguments:
+        ---------
+
+        X : np.ndarray of shape = (n_instances (n), n_dimensions (d), series_length (m))
```
changing to call them channels

TonyBagnall commented on sktime/classification/deep_learning/inception_time.py:
```diff
@@ -0,0 +1,673 @@
+        ----------
+        input_shape : tuple
+            The shape of the data fed into the input layer
+        nb_classes: int
```
n_classes

TonyBagnall commented on sktime/classification/deep_learning/resnet.py:
```diff
@@ -21,27 +21,54 @@ class ResNetClassifier(BaseDeepClassifier):
+            the convolution layers in one residual block, if not
+            a list, the same kernel size is used in all convolution layers
+
+        nb_epochs                   : int, default = 1500
```
we are going with n_epochs I think

TonyBagnall reviewed (changes requested):
so minor things, looking good

hadifawaz1999 commented:
Thanks @TonyBagnall , I fixed all the problems you mentioned, but for the example I had an issue with code quality (line too long when importing inceptionTime classifier) so I had to make it on two lines. Let me know if everything is good.

hadifawaz1999 commented:
If i do git pull main while in my branch locally wont it remove my changes ?

MatthewMiddlehurst commented:
It shouldn't, no. There is a single conflict that will have to be resolved but that can be done by deleting `lstmfcn.py` again.

![image](https://user-images.githubusercontent.com/25731235/223501036-c32eabc2-68f8-473f-83ce-e183d3cfc14b.png)

MatthewMiddlehurst commented:
Most of the changes will be merged automatically, if there is a local change from you and a change on main on the same file it should prompt you to resolve it and commit again before pushing.

TonyBagnall commented:
File /opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/sktime/classification/base.py:192, in BaseClassifier.fit(self, X, y)
    187         raise AttributeError(
    188             "self.n_jobs must be set if capability:multithreading is True"
    189         )
    191 # pass coerced and checked data to inner _fit
--> 192 self._fit(X,y)
    193 self.fit_time_ = int(round(time.time() * 1000)) - start
    195 # this should happen last

I dont think you want to set multithreaded to be true, it will run automatically on the GPU, n_jobs is for CPU threading

MatthewMiddlehurst commented on sktime/classification/deep_learning/inception_time.py:
```diff
@@ -0,0 +1,682 @@
+        probs = probs / self.n_classifiers
+
+        return probs
+
+    def get_test_params(cls, parameter_set="default"):
+        """Return testing parameter settings for the estimator.
```
We are missing `@classmethod` at the top of this method.

hadifawaz1999 replied:
Yeah I pushed now the fixed code for that in both InceptionTimeClassifier and IndividualInceptionClassifier methods

MatthewMiddlehurst commented on 02a_classification_multivariate_cnn.ipynb:
```diff
@@ -0,0 +1,140 @@
+     "is_executing": true
+    }
+   },
+   "outputs": [],
+   "source": [
+    "param_grid = {\"kernel_size\": [7, 9], \"avg_pool_size\": [3, 5]}\n",
+    "grid = GridSearchCV(network, param_grid=param_grid, cv=5)\n",
+    "grid.fit(X_train, y_train)\n",
```
CNN has `kernel_sizes` but call is `kernel_size`.

hadifawaz1999 replied:
oh yeah most of the architectures now use kernel_sizes, I can change it back to kernel_size if its best for everyone, I made it with 's' because its more than one, but I think its okay without the 's' :) what do you think @MatthewMiddlehurst @TonyBagnall ?

TonyBagnall replied:
I think best use the same for all, I dont mind which, if you choose sizes just change the notebook

hadifawaz1999 replied:
I changed everything back to kernel_size

MatthewMiddlehurst reviewed (commented):
Think these are the test failures.

MatthewMiddlehurst reviewed (commented):
Test parameter in CNN needs to be changed to `n_layers` from `n_conv_layers` I believe.

MatthewMiddlehurst commented:
The classifier `get_test_params` had it has a parameter, but the class itself does not. This is breaking the tests for CNNClassifier

hadifawaz1999 commented:
@MatthewMiddlehurst I added the ```n_conv_layers``` parameter to the CNN classifier set by default to None to avoid test fails.

MatthewMiddlehurst commented:
Hello, I think the changes in https://github.com/hadifawaz1999/scikit-time/pull/1 should fix the tests. There is a bit of a mixup beween `n_layers` and `n_conv_layers`. 

Also, not so relevant with those changes, but `n_conv_layers=None` as an argument must have an accompanying `self.n_conv_layers=n_conv_layers` in the init if it were to be included. Tests dont like arguments which go nowhere apparently 🙂.

hadifawaz1999 commented:
Hello @MatthewMiddlehurst , I made the changes as you said, made ```n_conv_layers``` a private variable in both the CNNNetwork and CNNClassifier classes, lets hope it works now

hadifawaz1999 commented:
Oh yeah that's my bad i will remove it, it was something i was testing for environments problems. Will push the branch with removing it when the tests are done.

TonyBagnall commented:
ah this fail is known, we need to put InceptionTime on the excluded list. Go to tests/_config.py and add it here 
![image](https://user-images.githubusercontent.com/9594042/224081917-1438d86f-9e6a-48d6-9129-5ede18cc464b.png)

hadifawaz1999 commented:
on all classifiers the pickle problem exists @MatthewMiddlehurst @TonyBagnall , should we ignore it for all classifiers ?

MatthewMiddlehurst reviewed (approved):
I think this should be good (hopefully). A bit more of a pain than I thought it would be, but probably the hardest step 🙂. 

Another time it would be good to chip away at these tests individually and figure out A. what is breaking them and B. if it's something we actually can/want to fix. First 3 probably more difficult, last 2 I think easier.

- "test_fit_idempotent",
- "test_persistence_via_pickle"
- "test_save_estimators_to_file"
- "test_fit_does_not_overwrite_hyper_params"
- "test_methods_have_no_side_effects"

hadifawaz1999 commented:
I think a first step would be to adapt the tests parameters to the new ones. like the n_conv_layers in cnn

MatthewMiddlehurst commented:
We should rethink the testing parameters and what the parameters are called, yeah. Ideally we want them to be quick as possible, and we could possibly use the new parameters to help with that. As you are probably aware at this point, the tests take an age to run and one of our goals is to reduce that by as much as possible 🙂. The issue is finding parameter values which cover as much as possible while remaining fast to run.

This issue we ran into in this PR was we renamed `n_conv_layers`, but kept that name in `get_test_parameters` i.e.

>         param1 = {
>             "n_epochs": 10,
>             "batch_size": 4,
>             "avg_pool_size": 4,
>             "n_conv_layers": 1,
>         }

So in the test suite, it would try to create a `CNNClassifier` and set `n_conv_layers` to 1, but would raise an error because `n_conv_layers` doesn't exist (as it was renamed).
